### ISSUE TITLE
[docs] - Setting headers on document after prod build

### ISSUE THREAD

**Issue by Sparkenstein** [tauri-apps/tauri#3917]
I am working with libvips to compile it to wasm with emscripten. the wasm module  needs access to sharedarraybuffer API, which can be only enabled by sending `Cross-Origin-Embedder-Policy` and `Cross-Origin-Opener-Policy` headers from document with values `require-corp` and `same-origin` respectively. it works just fine with `vite` because in dev server I can set any headers.

but in prod mode there's no 'server' that I can found in doc. how do I enable SharedArrayBuffer in prod mode? any other way to set headers? 

In my one observation I found all request in prod are going to `https://tauri.localhost` is that a local server?

**Comment by Sparkenstein** [comment]
maybe related: #1522

**Comment by nothingismagick** [comment]
`https://tauri.localhost` Is actually a "custom protocol" - so not really a server in the sense that other apps can just listen in on some port over loopback.

But as to the proper solution I am not sure ATM. You could always ship a server along with core (and set its headers appropriately). Not sure though that this would solve your problem elegantly.

**Comment by nothingismagick** [comment]
Also, it would be helpful to track the system you are using. Can you please add the results of `tauri info` ?

**Comment by lucasfernog** [comment]
Unfortunately there's no way to set response headers in Linux yet :(

**Comment by lucasfernog** [comment]
If you're using Windows or macOS, you can set custom headers using [on_web_resource_request](https://docs.rs/tauri/1.0.0-rc.6/tauri/window/struct.WindowBuilder.html#method.on_web_resource_request).

**Comment by lucasfernog** [comment]
Tested it and it still didn't work, see https://github.com/tauri-apps/tauri/issues/1522#issuecomment-1103394012

**Comment by Sparkenstein** [comment]
Sad. my full fledged app is ready and working but only in dev mode :P what a tragedy.

let me know if I can try something from my side, but I am not exactly proficient in rust. 

@nothingismagick  you mentioned I can try shipping my own server. How does that work? will it be equally safe? assuming it will open a tcp port on client machine.

**Comment by lucasfernog** [comment]
You can use a localhost server instead of the custom protocol with https://github.com/tauri-apps/tauri-plugin-localhost though it doesn't allow you to set custom headers yet.

**Comment by lucasfernog** [comment]
It does work on Windows though if you set the headers using on_web_resource_request `crossOriginIsolated` is `true` and `SharedArrayBuffer` is set.

**Comment by lucasfernog** [comment]
> You can use a localhost server instead of the custom protocol with [tauri-apps/tauri-plugin-localhost](https://github.com/tauri-apps/tauri-plugin-localhost) though it doesn't allow you to set custom headers yet.

Just updated it to allow you to set the headers (it works on macOS with the localhost server).

**Comment by Sparkenstein** [comment]
I am looking into this over this weekend, few questions I had
- what are the security implications of using localhost server? I assume it opens a port on client machine.
- using [on_web_resource_request](https://docs.rs/tauri/1.0.0-rc.6/tauri/window/struct.WindowBuilder.html#method.on_web_resource_request) will not need localhost plugin but then if I understand correctly app will not work on linux. is that tauri limitation or platform webview limitation? if it's tauri limitation any plan to resolve it?
- is localhost plugin guaranteed to work on all platforms? assuming the security implications are not that huge.

Edit: I tried both, both are failing with error
```
thread 'main' panicked at 'error while running tauri application: Setup(SetupError(WindowLabelAlreadyExists("main")))', src\main.rs:34:6
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: process didn't exit successfully: `target\debug\app.exe` (exit code: 101)dgetWin_0. Error = 0
error Command failed with exit code 101.
```

here's my main function:
```rust
fn main() {
  tauri::Builder::default()
    .setup(|app| {
      #[cfg(debug_assertions)]
      app.get_window("main").unwrap().open_devtools();

      WindowBuilder::new(app, "main", WindowUrl::App("index.html".into()))
        .on_web_resource_request(|req, resp| {
          if req.uri().starts_with("tauri://") {
            resp.headers_mut().insert(
              "Cross-Origin-Opener-Policy",
              "same-origin".try_into().unwrap(),
            );
            resp.headers_mut().insert(
              "Cross-Origin-Embedder-Policy",
              "require-corp".try_into().unwrap(),
            );
            println!("headers inserted");
          }
        })
        .build()?;
      Ok(())
    })
    .run(tauri::generate_context!())
    .expect("error while running tauri application");
}

```

**Comment by Sparkenstein** [comment]
I missed attaching my system information above, my bad. here's output of  `tauri info`

```
Environment
  › OS: Windows 10.0.19043 X64
  › Webview2: 101.0.1210.39
  › MSVC:
      - Visual Studio Build Tools 2019
  › Node.js: 16.13.1
  › npm: 8.1.2
  › pnpm: 6.27.1
  › yarn: 1.22.17
  › rustup: 1.24.3
  › rustc: 1.60.0
  › cargo: 1.60.0
  › Rust toolchain: stable-x86_64-pc-windows-msvc 

Packages
  › @tauri-apps/cli [NPM]: 1.0.0-rc.10
  › @tauri-apps/api [NPM]: 1.0.0-rc.5
  › tauri [RUST]: 1.0.0-rc.9,
  › tauri-build [RUST]: 1.0.0-rc.8,
  › tao [RUST]: 0.8.4,
  › wry [RUST]: 0.16.0,

App
  › build-type: bundle
  › CSP: default-src blob: data: filesystem: ws: wss: http: https: tauri: 'unsafe-eval' 'unsafe-inline' 'self' img-src: 'self'
  › distDir: ../dist
  › devPath: http://localhost:3000/
  › framework: React
  › bundler: Rollup

App directory structure
  ├─ .git
  ├─ .github
  ├─ .vscode
  ├─ dist
  ├─ node_modules
  ├─ public
  ├─ src
  ├─ src-tauri
  └─ vite
Done in 16.57s.
```

I am using yarn with `vite`

**Comment by lucasfernog** [comment]
1. what are the security implications of using localhost server? I assume it opens a port on client machine.
`Yeah your app assets will be accessible from other applications.`

2. using [on_web_resource_request](https://docs.rs/tauri/1.0.0-rc.6/tauri/window/struct.WindowBuilder.html#method.on_web_resource_request) will not need localhost plugin but then if I understand correctly app will not work on linux. is that tauri limitation or platform webview limitation? if it's tauri limitation any plan to resolve it?
`on_web_resource_request only works on the Tauri custom protocol (the default implementation for production apps - the ones you create using `tauri build`). It doesn't work on Linux yet because webkit2gtk only added custom headers support on version 2.36 and that release is not available in older distros`, see #3961

3. is localhost plugin guaranteed to work on all platforms? assuming the security implications are not that huge.
`It will work on all platforms.`

Your issue is actually that you're creating a second window with the `main` label, try changing `WindowBuilder::new(app, "main", WindowUrl::App("index.html".into()))` to `WindowBuilder::new(app, "second", WindowUrl::App("index.html".into())). The label must be unique.`

**Comment by lucasfernog** [comment]
Since we can set the headers on supported platforms, I'll close this one and track the crossOriginIsolated issue in #1522
